spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java

```diff
@@ -93,33 +95,51 @@ public TwoStepsConversions(CustomConversions customConversions, ObjectToKeyFacto
 
 		this.internalConversionService.addConverter(BYTE_ARRAY_TO_BLOB_CONVERTER);
 		this.internalConversionService.addConverter(BLOB_TO_BYTE_ARRAY_CONVERTER);
-
 	}
 
 	@Override
 	public <T> T convertOnRead(Object val, Class targetCollectionType,
 			Class targetComponentType) {
-		return convertOnRead(val, false, targetCollectionType, targetComponentType);
+		return (T) convertOnRead(val, false, false, false, targetCollectionType,
+				ClassTypeInformation.from(targetComponentType));
 	}
 
 	@Override
 	public <T> T convertOnReadEmbedded(Object val, Class targetCollectionType,
-			Class targetComponentType) {
-		return convertOnRead(val, true, targetCollectionType, targetComponentType);
+			Class targetComponentType, boolean isEmbeddedComponents) {
+		return (T) convertOnRead(val, !isEmbeddedComponents, false, isEmbeddedComponents,
+				targetCollectionType,
+				ClassTypeInformation.from(targetComponentType));
+	}
+
+	@Override
+	public <T> T convertOnReadEmbeddedMap(BaseEntity val, Class targetCollectionType,
+			TypeInformation targetComponentType, boolean embeddedMapValueIsEmbedded) {
+		return convertOnRead(val, true, true, embeddedMapValueIsEmbedded, null,
+				targetComponentType);
 	}
 
 	@SuppressWarnings("unchecked")
-	private <T> T convertOnRead(Object val, boolean isEmbedded,
-			Class targetCollectionType, Class targetComponentType) {
+	private <T> T convertOnRead(Object val, boolean isEmbedded, boolean isEmbeddedMap,
```

Comment:
Instead of the 3 booleans let's make an enum. 

also write a table to spell out the combinations in the ref doc.